ruby/ext/google/protobuf_c/message.c

```diff
@@ -139,7 +139,14 @@ int Message_initialize_kwarg(VALUE key, VALUE val, VALUE _self) {
              "Unknown field name in initialization map entry.");
   }
 
-  if (upb_fielddef_label(f) == UPB_LABEL_REPEATED) {
+  if (is_map_field(f)) {
+    if (TYPE(val) != T_HASH) {
+      rb_raise(rb_eArgError,
+               "Expected hashmap as initializer value for map field.");
```

Comment:
Your last point (copy vs. reference semantics) is I think a reasonable argument for not supporting this case, although arguably, if the user explicitly passes a `RepeatedField` or `Map`, they know what they're doing. For simplicity I'll keep the uniform all-ctor-args-are-copied-from-native-Ruby-types for now and we can always add this support later.